torchao/experimental/kernels/cpu/aarch64/kleidi/kai_matmul_clamp_f32_qai8dxp_qsi4c32p.h

```diff
@@ -0,0 +1,144 @@
+// Copyright (c) Meta Platforms, Inc. and affiliates.
+// All rights reserved.
+//
+// This source code is licensed under the license found in the
+// LICENSE file in the root directory of this source tree.
+
+#pragma once
+
+#include <cassert>
+#include <cstddef>
+#include <cstdint>
+#include <cstring>
+#include <limits>
+#include <vector>
+
+#include <kai/kai_common.h>
+#include <kai/ukernels/matmul/matmul_clamp_f32_qai8dxp_qsi4c32p/kai_matmul_clamp_f32_qai8dxp_qsi4c32p_interface.h>
+
+#include <torchao/experimental/kernels/cpu/aarch64/kleidi/pack.h>
+
+namespace torchao::kernels::cpu::aarch64::kleidi {
+
+// Helper functions
+// TODO: find a better place for these?
+
+size_t roundup(size_t a, size_t b) {
+  return ((a + b - 1) / b) * b;
+}
+
+uint16_t get_bf16_from_float(float f) {
+  uint16_t bf16;
+#if __BYTE_ORDER__ == __ORDER_BIG_ENDIAN__
+  memcpy(&bf16, &f, sizeof(uint16_t));
+#else
+  const void* fp = reinterpret_cast<const void*>(
+      reinterpret_cast<uintptr_t>(&f) + sizeof(float) - sizeof(uint16_t));
+  memcpy(&bf16, fp, sizeof(uint16_t));
+#endif // __BYTE_ORDER__
+  return bf16;
+}
+
+namespace kai_matmul_clamp_f32_qai8dxp_qsi4c32p {
+
+using Ukernel = struct kai_matmul_clamp_f32_qai8dxp_qsi4c32p_ukernel;
+
+int activation_data_size(const Ukernel ukernel, int m, int k) {
+  auto lhs_packing = get_lhs_packing();
+  return lhs_packing.get_lhs_packed_size(
+      m, k, ukernel.get_mr(), ukernel.get_kr(), ukernel.get_sr());
+}
+
+void prepare_activation_data(
+    const Ukernel ukernel,
+    void* activation_data,
+    int m,
+    int k,
+    const float* activations) {
+  auto lhs_pack = get_lhs_packing();
+
+  lhs_pack.run_lhs_pack(
+      m,
+      k,
+      ukernel.get_mr(),
+      ukernel.get_kr(),
+      ukernel.get_sr(),
+      /*m_index_start=*/0,
+      activations,
+      /*lhs_stride=*/k * sizeof(float),
+      activation_data);
+}
+
+int weight_data_size(const Ukernel ukernel, int n, int k, int group_size) {
+  auto rhs_pack = get_rhs_packing();
+  return rhs_pack.get_rhs_packed_size(
+      n,
+      k,
+      ukernel.get_nr(),
+      ukernel.get_kr(),
+      ukernel.get_sr(),
+      group_size,
+      kai_datatype::kai_dt_bf16);
```

Comment:
why bf16?

Reply:
I think the kleidi kernel keeps scales as bf16 to save space.

Reply:
yeah we asked Kleidi to do this :p